### PR TITLE
Bug 1825831: Refactoring collector, add Doc and doc generator

### PR DESCRIPTION
**Pull request by martinkunc**
The goal here is to
- Refactor the clusterconfig.Gather and extract the methods out for better separation and testing
- Document the individual methods in code, add description, and location in the archive
- Generate documentation from the methods. It collects Gather... methods and if exists complementary Example... methods to collect the anonymized output example.
- Resulting documentation is placed in docs/gathered-data.md

### CODE REVIEW

**Comment by alexandrevicenzi** [comment]
@martinkunc is there any BZ or Jira related to this PR?

File: pkg/gather/clusterconfig/clusterconfig.go [hunk 1]
```diff
@@ -65,208 +65,376 @@ var reInvalidUIDCharacter = regexp.MustCompile(`[^a-z0-9\-]`)
...
+// GatherMostRecentMetrics gathers cluster Federated Monitoring metrics
+//
+// The GET REST query to URL /federate
+// Gathered metrics:
+//   ALERTS
+//   etcd_object_counts
+//   cluster_installer
+//
+// Location of operators in archive: config/metrics/
+// see: docs/insights-archive-sample/config/metrics
```
Comment (alexandrevicenzi) [inline pkg/gather/clusterconfig/clusterconfig.go:94]: See could be Title case

File: pkg/gather/clusterconfig/clusterconfig.go [hunk 2]
```diff
@@ -65,208 +65,376 @@ var reInvalidUIDCharacter = regexp.MustCompile(`[^a-z0-9\-]`)
...
+		GatherClusterIngress(i),
+		GatherClusterProxy(i),
+		GatherCertificateSigningRequests(i),
+	)
+}
 
-			now := time.Now()
-			for _, item := range config.Items {
-				if isHealthyOperator(&item) {
+// GatherMostRecentMetrics gathers cluster Federated Monitoring metrics
```
Comment (alexandrevicenzi) [inline pkg/gather/clusterconfig/clusterconfig.go:85]: dot at the end would be nice

File: pkg/gather/clusterconfig/clusterconfig.go [hunk 3]
```diff
@@ -65,208 +65,376 @@ var reInvalidUIDCharacter = regexp.MustCompile(`[^a-z0-9\-]`)
...
+//
+// The GET REST query to URL /federate
```
Comment (alexandrevicenzi) [inline pkg/gather/clusterconfig/clusterconfig.go:87]: Sounds a bit better:

Same data as GET /federate

File: pkg/gather/clusterconfig/clusterconfig.go [hunk 4]
```diff
@@ -65,208 +65,376 @@ var reInvalidUIDCharacter = regexp.MustCompile(`[^a-z0-9\-]`)
...
+	}
+}
+
+// GatherClusterOperators collects all ClusterOperators
+// In unhealthy operators it finds unhealthy Pods
+//
+// The Kubernetes api https://github.com/openshift/client-go/blob/master/config/clientset/versioned/typed/config/v1/clusteroperator.go#L62
+// Response see https://docs.openshift.com/container-platform/4.3/rest_api/index.html#clusteroperatorlist-v1config-openshift-io
+//
+// Location of operators in archive: config/clusteroperator/
```
Comment (alexandrevicenzi) [inline pkg/gather/clusterconfig/clusterconfig.go:124]: One function creates two archives? Should we split this function?

File: pkg/gather/clusterconfig/clusterconfig.go [hunk 5]
```diff
@@ -65,208 +65,376 @@ var reInvalidUIDCharacter = regexp.MustCompile(`[^a-z0-9\-]`)
...
+// see: docs/insights-archive-sample/config/clusteroperator
+// Location of pods in archive: config/pod/
```
Comment (alexandrevicenzi) [inline pkg/gather/clusterconfig/clusterconfig.go:126]: No example for pods?

File: pkg/gather/clusterconfig/clusterconfig.go [hunk 6]
```diff
@@ -65,208 +65,376 @@ var reInvalidUIDCharacter = regexp.MustCompile(`[^a-z0-9\-]`)
...
+//
+// The node is unhealthy when:
+// the operator.Status.Conditions.Condition Type
+//   is OperatorDegrated and Status is True or
+//      OperatorAvailable and Status is False
+//
+// The Kubernetes api https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/node.go#L78
+// Response see https://docs.openshift.com/container-platform/4.3/rest_api/index.html#nodelist-v1core
+//
+// Location in archive: config/node/
```
Comment (alexandrevicenzi) [inline pkg/gather/clusterconfig/clusterconfig.go:184]: No See?

File: docs/gathered-data.md [hunk 7]
```diff
@@ -0,0 +1,188 @@
+## ClusterIngress
```
Comment (alexandrevicenzi) [inline docs/gathered-data.md:1]: No note that his is autogen file? Someone can manually update

File: cmd/gendoc/main.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
+package main
```
Comment (alexandrevicenzi) [inline cmd/gendoc/main.go:1]: Why do we need gendoc when we have goland godoc?

I don't see much value in exporting to MD when we already have HTML.

File: pkg/gather/clusterconfig/clusterconfig.go (hunk 5 above)
Comment (martinkunc) [reply pkg/gather/clusterconfig/clusterconfig.go:126]: I would like to add more samples later. I just have few samples ready and it takes some time to get each prepared with the test.

File: cmd/gendoc/main.go (hunk 8 above)
Comment (martinkunc) [reply cmd/gendoc/main.go:1]: The generator also generated Examples, and output data size. And there are some more ideas how to make whole effort easier with the specific generator.

Comment (alexandrevicenzi) [reply cmd/gendoc/main.go:1]: Ok

**Comment by martinkunc** [comment]
/retitle Bug 1825831: Refactoring collector, add Doc and doc generator

File: README.md [hunk 9]
```diff
@@ -8,6 +8,10 @@ This cluster operator gathers anonymized system configuration and reports it to
 * ClusterOperator objects
 * All non-secret global config (hostnames and URLs anonymized)
 
+The list of all collected data with description, location in produced archive and link to Api and some examples is at [docs/gathered-data.md](docs/gathered-data.md)
+
+The resulted data are packed in .tar.gz archive with folder structure indicated in the document. Example of such archive is at [docs/insights-archive-sample](docs/insights-archive-sample).
```
Comment (tremes) [inline README.md:13]: `The resulted data is` maybe `The resulting data is` ...

File: docs/gathered-data.md [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+Response see https://docs.openshift.com/container-platform/4.3/rest_api/index.html#oauth-v1-config-openshift-io
+
+Location in archive: config/oauth/
+See: docs/insights-archive-sample/config/oauth
+
+
+## ClusterOperators
+
+collects all ClusterOperators.
+In unhealthy operators it finds unhealthy Pods
```
Comment (tremes) [inline docs/gathered-data.md:94]: What about `It finds unhealthy Pods for unhealthy operators` ?

**Pull request merged**